### PR TITLE
[scene_host] Add Views V2 support

### PR DESCRIPTION
**Pull request by arbreng**
This allows using the new Views API from Flutter programs on Fuchsia.

### CODE REVIEW

File: flow/scene_update_context.cc [hunk 1]
```diff
@@ -356,7 +375,7 @@ SceneUpdateContext::Frame::Frame(SceneUpdateContext& context,
       paint_bounds_(SkRect::MakeEmpty()),
       layer_(layer) {
   if (elevation != 0.0)
-    entity_node().SetTranslationRH(0.f, 0.f, -elevation);
+    entity_node().SetTranslationRH(0.f, 0.f, elevation);
```
Comment (schwa423) [inline flow/scene_update_context.cc]: Should probably be -elevation (why the change?)

Note: this would have been "elevation" when Scenic was left-handed, but needs to be "-elevation" now that it's right-handed.

File: flow/export_node.h [hunk 2]
```diff
@@ -2,8 +2,8 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-#ifndef FLUTTER_FLOW_EXPORT_NODE_H_
-#define FLUTTER_FLOW_EXPORT_NODE_H_
+#ifndef THIRD_PARTY_FLUTTER_FLOW_EXPORT_NODE_H_
```
Comment (schwa423) [inline flow/export_node.h]: Revert these erroneous header-guard changes.

File: flow/layers/child_scene_layer.cc [hunk 3]
```diff
@@ -21,14 +21,18 @@ void ChildSceneLayer::Paint(PaintContext& context) const {
 void ChildSceneLayer::UpdateScene(SceneUpdateContext& context) {
   FML_DCHECK(needs_system_composite());
 
-  // TODO(MZ-191): Set clip.
+  // TODO(SCN-191): Set clip.
   // It's worth asking whether all children should be clipped implicitly
   // or whether we should leave this up to the Flutter application to decide.
   // In some situations, it might be useful to allow children to draw
   // outside of their layout bounds.
+  FML_LOG(ERROR) << "ChildSceneLayer::UpdateScene";
```
Comment (schwa423) [inline flow/layers/child_scene_layer.cc]: DCHECK that at most one of the two is non-null?

Or, just say:

if (view_holder_proxy_) instead of "else if".  The former is probably better, if that's the expectation.

File: flow/scene_update_context.cc [hunk 4]
```diff
@@ -7,38 +7,12 @@
...
-}
-
-void SceneUpdateContext::AddExportNode(ExportNode* export_node) {
-  export_nodes_.insert(export_node);  // Might already have been added.
-}
-
-void SceneUpdateContext::RemoveExportNode(ExportNode* export_node) {
-  export_nodes_.erase(export_node);
-}
+namespace {
```
Comment (schwa423) [inline flow/scene_update_context.cc]: Anonymous namespace unnecessary.  The 'static' in front of SetEntityNodeClipPlanes() is equivalent to putting it in an anonymous namespace (restricts scope to the compilation unit).

File: flow/scene_update_context.h [hunk 5]
```diff
@@ -148,18 +150,26 @@ class SceneUpdateContext {
...
+  // |export_node| is destroyed.
   void AddChildScene(ExportNode* export_node,
                      SkPoint offset,
                      bool hit_testable);
 
-  // Adds reference to |export_node| so we can call export_node->Dispose() in
-  // our destructor. Caller is responsible for calling RemoveExportNode() before
-  // |export_node| is destroyed.
-  void AddExportNode(ExportNode* export_node);
+  // Adds reference to |view)_holder| so we can call view_holder->Dispose() in
```
Comment (schwa423) [inline flow/scene_update_context.h]: Typo |view)_holder|

File: flow/view_holder.cc [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  // Ensure that we properly released the resource.
+  FML_DCHECK(!view_holder_);
+  FML_DCHECK(scene_update_context_ == nullptr);
+}
+
+void ViewHolder::Bind(SceneUpdateContext& context,
+                      scenic::ContainerNode& container,
+                      const SkPoint& offset,
+                      bool hit_testable) {
+  FML_LOG(ERROR) << "ViewHolder::Bind";
```
Comment (schwa423) [inline flow/view_holder.cc]: Remove before submit.

File: flow/view_holder.cc [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  if (view_holder_token_.value) {
+    FML_LOG(ERROR) << "CREATE NODES";
```
Comment (schwa423) [inline flow/view_holder.cc]: Remove before submit.

File: flow/view_holder.cc [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    // Happens first time we bind.
+    // The Resource ID message MUST be posted before the resource is created, or
+    // race conditions could occur!
+    scenic::ResourceId view_holder_id = container.session()->next_resource_id();
+    on_bind_callback_(view_holder_id);
+
+    // Create resources.  This might cause an immediate session flush, so this
+    // must be done after registering the view_holder_id via the callback.
```
Comment (schwa423) [inline flow/view_holder.cc]: Even if it caused a session flush, this isn't necessary for two reasons:
1) even if Scenic sent out events, the event loop handler won't run until the current call-stack is finished.
2) note that 1) might not be true if the listener event-loop is on a different thread.  Nevertheless, flushed events won't be applied until Present() is called.

File: flow/view_holder.cc [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        container.session(), std::move(view_holder_token_.value),
+        "Flutter SceneHost"));
+    transform_node_.reset(new scenic::EntityNode(container.session()));
+    transform_node_->Attach(*view_holder_);
+
+    scene_update_context_ = &context;
+  }
+
+  if (view_holder_) {
+    FML_LOG(ERROR) << "ATTACH NODES";
```
Comment (schwa423) [inline flow/view_holder.cc]: Remove before submit.

File: flow/view_holder.cc [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    transform_node_->SetTranslation(offset.x(), offset.y(), 0.f);
```
Comment (schwa423) [inline flow/view_holder.cc]: We want to add a fixed elevation to avoid z-fighting with children.  The correct value is TBD, but 1.0 is probably a good starting point.

File: flow/view_holder.cc [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        std::make_unique<fuchsia::ui::gfx::ViewProperties>(new_properties);
+    return;
+  }
+
+  // TODO: focus change
+  FML_LOG(ERROR) << "ViewHolder::SetProperties";
+  view_holder_->SetViewProperties(new_properties);
+}
+
+ViewHolderProxy::ViewHolderProxy(
```
Comment (schwa423) [inline flow/view_holder.cc]: Are there other things in Flutter named Holder besides ExportNodeHolder.  If this is a common naming pattern, it's probably better to say ViewHolderHolder rather than to "innovate" with this naming.  Or perhaps rename "ViewHolder" to "EmbeddedView".

File: flow/view_holder.cc (hunk 8 above)
Comment (arbreng) [reply flow/view_holder.cc]: Nice, good point.  This lets me avoid using next_resource_id()

File: flow/scene_update_context.cc (hunk 1 above)
Comment (arbreng) [reply flow/scene_update_context.cc]: I think a bad merge might have caused this.  I started this change when scenic was left-handed, then jiri updated at one point and scenic became right-handed.

This ended up being the source of the disappearing views bug.

File: flow/export_node.h (hunk 2 above)
Comment (arbreng) [reply flow/export_node.h]: Done

File: flow/scene_update_context.cc (hunk 4 above)
Comment (arbreng) [reply flow/scene_update_context.cc]: Done

File: flow/view_holder.cc (hunk 10 above)
Comment (arbreng) [reply flow/view_holder.cc]: I added -0.1, that is what we have been using everywhere else

File: flow/layers/child_scene_layer.cc (hunk 3 above)
Comment (arbreng) [reply flow/layers/child_scene_layer.cc]: Done

File: flow/view_holder.cc (hunk 11 above)
Comment (arbreng) [reply flow/view_holder.cc]: I ended up refactoring the code to remove this split entirely, so the issue with the naming went away too

File: lib/ui/compositing/scene_host.cc [hunk 12]
```diff
@@ -9,8 +9,11 @@
 #include "third_party/tonic/dart_binding_macros.h"
 #include "third_party/tonic/dart_library_natives.h"
 
-#ifdef OS_FUCHSIA
-#include "dart-pkg/zircon/sdk_ext/handle.h"
+#if defined(OS_FUCHSIA)
```
Comment (chinmaygarde) [inline lib/ui/compositing/scene_host.cc]: I thought this TU was Fuchsia only, can we update the build rules to compile this just on Fuchsia and remove the preprocessor guards?

File: lib/ui/window/window.h [hunk 13]
```diff
@@ -85,6 +99,10 @@ class Window final {
   WindowClient* client_;
   tonic::DartPersistentValue library_;
   ViewportMetrics viewport_metrics_;
+#if defined(OS_FUCHSIA)
```
Comment (chinmaygarde) [inline lib/ui/window/window.h]: Since we are trying to remove Fuchsia specific code in the engine, this can be stored on the engine or platform view subclass in the Flutter runner on Fuchsia.

File: lib/ui/window/window.cc [hunk 14]
```diff
@@ -249,6 +286,33 @@ void Window::UpdateAccessibilityFeatures(int32_t values) {
 }
 
 void Window::DispatchPlatformMessage(fml::RefPtr<PlatformMessage> message) {
+#if defined(OS_FUCHSIA)
```
Comment (chinmaygarde) [inline lib/ui/window/window.cc]: This can be handled in the platform view subclass in the Flutter runner. This would remove the need to have Fuchsia specific code in the engine.

File: lib/ui/compositing/scene_host.h [hunk 15]
```diff
@@ -11,7 +11,11 @@
 #include "flutter/lib/ui/dart_wrapper.h"
 
 #if defined(OS_FUCHSIA)
+#include "dart-pkg/zircon/sdk_ext/handle.h"
```
Comment (chinmaygarde) [inline lib/ui/compositing/scene_host.h]: We should just make this TU and its headers Fuchsia specific. They are not used on any other platform.

File: lib/ui/compositing/scene_host.cc [hunk 16]
```diff
@@ -19,40 +22,174 @@ static void SceneHost_constructor(Dart_NativeArguments args) {
...
+  if (view_holder_) {
+    view_holder_->SetProperties(width, height, insetTop, insetRight,
+        insetBottom, insetLeft, focusable);
+  }
+#endif
+}
+
 }  // namespace blink
+
```
Comment (chinmaygarde) [inline lib/ui/compositing/scene_host.cc]: Missing newline.

File: lib/ui/compositing/scene_host.cc (hunk 12 above)
Comment (arbreng) [reply lib/ui/compositing/scene_host.cc]: OK, SGTM.

What happens if someone tries to use a SceneHost on another platform?  Right now they can create one, but it will be useless.  After this change, I think even creating one will produce an error because the native hook is unimplemented.  The documentation *does* say "Fuchsia-only" for this class though.  Is all of this OK?

File: lib/ui/compositing/scene_host.cc (hunk 16 above)
Comment (arbreng) [reply lib/ui/compositing/scene_host.cc]: Done

File: lib/ui/compositing/scene_host.h (hunk 15 above)
Comment (arbreng) [reply lib/ui/compositing/scene_host.h]: Done

File: lib/ui/window/window.cc (hunk 14 above)
Comment (arbreng) [reply lib/ui/window/window.cc]: Nice!  Yeah, I was kinda unclear on which object owned which so this was a naive attempt.  I'll move it all to PlatformView in flutter_runner

File: lib/ui/window/window.h (hunk 13 above)
Comment (arbreng) [reply lib/ui/window/window.h]: Done

Comment (arbreng) [reply lib/ui/window/window.h]: I wasn't able to move this into PlatformView, but I did move it into SceneHost which is Fuchsia-specific

File: lib/ui/window/window.cc (hunk 14 above)
Comment (arbreng) [reply lib/ui/window/window.cc]: Fixed by using thread locals in SceneHost
